rfcs/20190910-struct-tensor.md

```diff
@@ -0,0 +1,885 @@
+information about a user and a recipe:
+
+```python
+struct = tf.struct.constant({
+  'user_embedding': [0.8, 2.1, 0.3, 0.1, 9.2, 1.8],
+  'recipe': {
+    'title': 'Snickerdoodle cookies',
+    'est_time': 55.0,
+    'ingredients': [
+        {'amount': 3.0, 'unit': 'cup', 'name': 'flour'},
+        {'amount': 1.0, 'unit': 'cup', 'name': 'white sugar'},
+        {'amount': 0.5, 'unit': 'cup', 'name': 'brown sugar'},
+        {'amount': 1.0, 'unit': 'cup', 'name': 'butter'},
+        {'amount': 1.0, 'unit': 'teaspoon', 'name': 'cinnamon'},
+        {'amount': 2.0, 'unit': 'teaspoon', 'name': 'cream of tartar'},
+         ...],
+    'step': [
+         {'description': 'Preheat oven to 350 degrees F.'}
+         {'description': 'Whisk together the flour, cinnamon, baking soda, salt, ...'}
+         {'description': 'Using an electric mixer, cream together the butter and sugar ...'}
+         {'description': 'Add the dry ingredients to the wet ingredients and mix until ...'}
+         ...],
+    'user_rating': [
+         {'user_embedding': [0.7, 2.0, 0.3, 0.3, 5.2, 2.2], score: 0.8},
+         {'user_embedding': [1.4, 0.0, 3.1, 1.1, 1.2, 0.3], score: 0.4},
+         ...],
+   })
+```
+
+In this example:
+
+* The root object `struct` is a scalar `StructTensor` (i.e., a single
+  structure).
+* The `recipe` field contains a scalar `StructTensor` (a single nested
+  structure).
+* The nested fields `ingredients`, `step`, and `user_rating` contain vector
+  `StructTensor`s (i.e. 1D collections of structures with the same schema).
+* The nested fields `title`, `est_time`, `amount`, `unit`, `name`,
+  `description`, and `score` contain scalar `Tensor`s.
+* The nested `user_embedding` field contains a vector `Tensor`.
+
+In the initial implementation, the value for a field may be a `Tensor`, a
+`RaggedTensor`, or a nested `StructTensor`, where nested `StructTensor`s could
+contain single structures (scalar `StructTensor`) or collections of structures
+(non-scalar `StructTensor`).  In the future, we may add support for additional
+value types, such as `SparseTensor`s.
+
+A `StructTensor`'s schema constrains the type for each field:
+
+* **For fields with `Tensor` values**, the schema specifies the field value's
+  `dtype` and `shape`.
+* **For fields with `RaggedTensor` values**, the schema specifies the field
+  value's `dtype` and `ragged_rank`.
+* **For fields with `StructTensor` values**, the schema specifies the nested
+  `StructTensor` value's `shape` (e.g., a single structure vs a list of
+  structures), and the name and type for each nested field.
+
+For multidimensional `StructTensor` (such as a vector of structures), the
+individual structures must all share the same schema.
+
+### Usage
+
+#### Construction
+
+`StructTensor`s can be constructed from a variety of sources, including:
+
+* Nested Python dictionaries and lists.
+* Serialized protocol buffer messages.
+* JSON-encoded protocol buffers.
+* Apache Arrow record batches.
```

Comment:
You might consider accepting an Arrow Table, https://arrow.apache.org/docs/python/data.html#tables. This is currently only part of C++\Python Arrow, but is a higher-level structure that offers some benefits over a RecordBatch when working with Arrow data in Python. A RecordBatch is a standard data structure across Arrow, so it isn't a bad choice either.

Reply:
I think we should be able to support both (though tables may not be zero-copy)